### PR TITLE
Multiplex requests over a single subscription

### PR DESCRIPTION
**Pull request by caspervonb**

### CODE REVIEW

**Comment by caspervonb** [comment]
So we might have to take out inbox on request builder. Or keep legacy requests in for just that scenario.

**Comment by caspervonb** [comment]
Some benches against main

```

     Running benches/core_nats.rs (target/release/deps/core_nats-9f9b494bfb99583b)
Gnuplot not found, using plotters backend
Benchmarking request amount/32: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 6.5s.
request amount/32       time:   [673.88 ms 682.44 ms 692.60 ms]
                        thrpt:  [14.438 Kelem/s 14.653 Kelem/s 14.840 Kelem/s]
                 change:
                        time:   [-9.2688% -7.6984% -5.9234%] (p = 0.00 < 0.05)
                        thrpt:  [+6.2964% +8.3405% +10.216%]
                        Performance has improved.
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) low mild
  1 (10.00%) high severe
Benchmarking request amount/1024: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 6.6s.
request amount/1024     time:   [686.72 ms 691.14 ms 695.54 ms]
                        thrpt:  [14.377 Kelem/s 14.469 Kelem/s 14.562 Kelem/s]
                 change:
                        time:   [-9.2669% -8.2935% -7.5109%] (p = 0.00 < 0.05)
                        thrpt:  [+8.1208% +9.0435% +10.213%]
                        Performance has improved.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) low mild
Benchmarking request amount/8192: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 7.4s.
request amount/8192     time:   [769.48 ms 780.60 ms 797.06 ms]
                        thrpt:  [12.546 Kelem/s 12.811 Kelem/s 12.996 Kelem/s]
                 change:
                        time:   [-7.0389% -4.8250% -2.2368%] (p = 0.00 < 0.05)
                        thrpt:  [+2.2879% +5.0696% +7.5719%]
                        Performance has improved.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high severe

     Running unittests src/lib.rs (target/release/deps/nats-2e00c16348cd9698)

```

File: async-nats/src/lib.rs [hunk 1]
```diff
@@ -484,6 +501,25 @@ impl ConnectionHandler {
...
+                                reply,
+                                payload,
+                                headers,
+                                status,
+                                description,
+                                length,
+                            };
+
+                            // TODO don't unwrap
+                            sender.send(message).unwrap();
```
Comment (Jarema) [inline async-nats/src/lib.rs:520]: Fix the unwrap.

**Review by Jarema — CHANGES_REQUESTED** [review]
Do the TODO :)

File: async-nats/src/lib.rs (hunk 1 above)
Comment (caspervonb) [reply async-nats/src/lib.rs:520]: Ups

File: async-nats/src/lib.rs [hunk 2]
```diff
@@ -484,6 +501,26 @@ impl ConnectionHandler {
                             self.handle_flush().await?;
                         }
                     }
+                } else if let Some(multiplexer) = self.multiplexer.as_mut() {
```
Comment (Jarema) [inline async-nats/src/lib.rs:504]: I would check if sid is as we expect, just to avoid any edge cases.

**Review by Jarema — CHANGES_REQUESTED** [review]
Looks good!

One change request.

**Pull request merged**